src/sse/sse_butterflies.rs

```diff
@@ -0,0 +1,4246 @@
+        *output_slice.add(0) = val[0];
+        *output_slice.add(1) = val[1];
+        *output_slice.add(2) = val[2];
+    }
+
+    // length 3 fft of x, given as x0, x1, x2.
+    // result is [X0, X1, X2]
+    #[inline(always)]
+    pub(crate) unsafe fn perform_fft_direct(
+        &self,
+        value0: __m128d,
+        value1: __m128d,
+        value2: __m128d,
+    ) -> [__m128d; 3] {
+        // This is a SSE translation of the scalar 3-point butterfly
+        let x12p = _mm_add_pd(value1, value2);
+        let x12n = _mm_sub_pd(value1, value2);
+        let sum = _mm_add_pd(value0, x12p);
+
+        let temp_a = _mm_mul_pd(self.twiddle1re, x12p);
+        let temp_a = _mm_add_pd(temp_a, value0);
+
+        let n_rot = self.rotate.rotate(x12n);
+        let temp_b = _mm_mul_pd(self.twiddle1im, n_rot);
+
+        let x1 = _mm_add_pd(temp_a, temp_b);
+        let x2 = _mm_sub_pd(temp_a, temp_b);
+        [sum, x1, x2]
+    }
+}
+
+//   _  _             _________  _     _ _
+//  | || |           |___ /___ \| |__ (_) |_
+//  | || |_   _____    |_ \ __) | '_ \| | __|
+//  |__   _| |_____|  ___) / __/| |_) | | |_
+//     |_|           |____/_____|_.__/|_|\__|
+//
+
+pub struct SseF32Butterfly4<T> {
+    direction: FftDirection,
+    _phantom: std::marker::PhantomData<T>,
+    rotate: Rotate90F32,
+}
+
+boilerplate_fft_sse_f32_butterfly!(SseF32Butterfly4, 4, |this: &SseF32Butterfly4<_>| this
+    .direction);
+boilerplate_fft_sse_common_butterfly!(SseF32Butterfly4, 4, |this: &SseF32Butterfly4<_>| this
+    .direction);
+impl<T: FftNum> SseF32Butterfly4<T> {
+    #[inline(always)]
+    pub fn new(direction: FftDirection) -> Self {
+        assert_f32::<T>();
+        let rotate = if direction == FftDirection::Inverse {
+            Rotate90F32::new(true)
+        } else {
+            Rotate90F32::new(false)
+        };
+        Self {
+            direction,
+            _phantom: std::marker::PhantomData,
+            rotate,
+        }
+    }
+    #[inline(always)]
+    pub(crate) unsafe fn perform_fft_contiguous(
+        &self,
+        input: RawSlice<Complex<T>>,
+        output: RawSliceMut<Complex<T>>,
+    ) {
+        let value01 = _mm_loadu_ps(input.as_ptr() as *const f32);
+        let value23 = _mm_loadu_ps(input.as_ptr().add(2) as *const f32);
+
+        let temp = self.perform_fft_direct(value01, value23);
+
+        let array = std::mem::transmute::<[__m128; 2], [Complex<f32>; 4]>(temp);
+
+        let output_slice = output.as_mut_ptr() as *mut Complex<f32>;
+
+        *output_slice.add(0) = array[0];
+        *output_slice.add(1) = array[1];
+        *output_slice.add(2) = array[2];
+        *output_slice.add(3) = array[3];
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn perform_dual_fft_contiguous(
+        &self,
+        input: RawSlice<Complex<T>>,
+        output: RawSliceMut<Complex<T>>,
+    ) {
+        let value01a = _mm_loadu_ps(input.as_ptr() as *const f32);
```

Comment:
My top priority request for this PR is to remove all raw pointer arithmetic from FFT code. Instead, take a look at how the AVX code does it: The `AvxArray` trait provides a method on `RawSlice` called `load_complex` - which does a debug assert and then handles all the pointer arithmetic internally.

I recommend using the same model here: Create a  `SseArray`  trait, implement it for RawSlice<Complex<f32>> and RawSlice<Complex<f64>>, and create  a `SseArrayMut` trait and implement it for RawSliceMut<Complex<f32>> and RawSliceMut<Complex<f64>>. Then, update as much FFT code as possible to use these new trait methods.

I caught countless of my own mistakes via that debug assert, and having it would make the SSE code substantially more safe..

Reply:
good points, will look into it!

Reply:
I added a SseVector(Mut) trait and updated one small part of the 4-point butterfly: https://github.com/HEnquist/RustFFT/blob/ssesimple/src/sse/sse_butterflies.rs#L349
This will get rid of some repetitive casting/transmuting and a lot of unsafe pointer stuff.
Is this what you meant?
For this, I also use the `new_transmuted` constructor for the RawSlices: https://github.com/HEnquist/RustFFT/blob/ssesimple/src/sse/sse_butterflies.rs#L23

Reply:
Yes, that's what I had in mind. I don have one request though: Call the trait `SseArray`. In the AVX module, `AvxVector` is implemented on `__m256` and `__m128` and `__m256d and `__m128d` and represents a generic AVX register. Separately, there's a `AvxArray` trait which is implemented on slices and RawSlice, In this case, we're making a trait that operates on RawSlice (and possibly on slices), so the matching name would be SseArray.

We won't have a SseVector for now. I'd like to create one eventually, because it would allow for writing a single generic version of SSE Radix4 instead of having separate f32 and f64 versions, but that doesn't need to happen in this PR.

Reply:
I renamed and started implementing for all the butterflies. I'm about halfway done with that, quite a few lines of code to update..